server/src/main/java/com/thoughtworks/go/server/service/plugins/AnalyticsPluginAssetsService.java

```diff
@@ -132,6 +138,26 @@ private void cacheStaticAssets(String pluginId) {
         }
     }
 
+    private void safeCopyExternalAssetsToPluginAssetRoot(final String pluginAssetsRoot) {
+        Path externalAssetsPath = Paths.get(systemEnvironment.get(SystemEnvironment.GO_ANALYTICS_PLUGIN_EXTERNAL_ASSETS));
+        if (externalAssetsPath == null || !Files.exists(externalAssetsPath) || !Files.isDirectory(externalAssetsPath)) {
+            LOGGER.debug("Analytics plugin external assets path ({}) does not exist or is not a directory. Not loading any assets.", externalAssetsPath);
+            return;
+        }
+
+        try {
+            Files.list(externalAssetsPath).forEach(path -> {
+                try {
+                    Files.copy(path, Paths.get(pluginAssetsRoot, path.getFileName().toString()));
+                } catch (Exception e) {
+                    LOGGER.error("Unable to copy analytics plugin external asset ({}) to plugin assets root.", path, e);
+                }
+            });
+        } catch (Exception e) {
+            LOGGER.error("Unable to list files in analytics plugin external assets location ({}).", externalAssetsPath, e);
```

Comment:
Yeah, they will not be visible directly, since this is _after_ the plugin has loaded. Maybe it'll show up in the plugin page.

The current approach logs it. Two problems with making it explicit:

1. We would be assuming all _future_ analytics plugins need these files. That would be wrong. This is just a generic mechanism to load external (JS) assets into the plugin iframe.

2. The commercial plugin does not need any external assets, since we can package highcharts along with it (as we have purchased licenses which can be used commercially).

Reply:
By the way, I can bubble up the error here, since this is more about an inability to read the file or directory (mostly permission issues). What I'd like to not do is to explicitly list the expected JS files. That would hamstring any future analytics plugins.

Reply:
I think that makes sense - keeping it generic and just validating the directory is readable.

Reply:
(which you are doing -- but lets allow the exception through now)

Reply:
Done. I'm just verifying it works fine.

Reply:
Looks ok. It shows this in the plugins page:

<img width="1182" alt="2020_06_04_10-24-53" src="https://user-images.githubusercontent.com/172235/83739577-b0601180-a64d-11ea-9b0d-87b5316881dd.png">

... and this in the logs:

```
2020-06-04 10:23:26,663 WARN  [Thread-77] DefaultPluginJarLocationMonitor:260 - Plugin listener failed
java.lang.RuntimeException: Failed to load plugin: plugins_work/gocd-analytics.jar
        at com.thoughtworks.go.plugin.infra.PluginLoader.loadPlugin(PluginLoader.java:76)
        at com.thoughtworks.go.plugin.infra.listeners.DefaultPluginJarChangeListener.refreshBundle(DefaultPluginJarChangeListener.java:180)
        at com.thoughtworks.go.plugin.infra.listeners.DefaultPluginJarChangeListener.addPlugin(DefaultPluginJarChangeListener.java:123)
        at com.thoughtworks.go.plugin.infra.listeners.DefaultPluginJarChangeListener.pluginJarAdded(DefaultPluginJarChangeListener.java:72)
        at com.thoughtworks.go.plugin.infra.monitor.DefaultPluginJarLocationMonitor$PluginLocationMonitorThread$DoOnAllListeners.lambda$pluginJarAdded$0(DefaultPluginJarLocationMonitor.java:237)
        at com.thoughtworks.go.plugin.infra.monitor.DefaultPluginJarLocationMonitor$PluginLocationMonitorThread$DoOnAllListeners.doOnAllPluginJarChangeListener(DefaultPluginJarLocationMonitor.java:258)
        at com.thoughtworks.go.plugin.infra.monitor.DefaultPluginJarLocationMonitor$PluginLocationMonitorThread$DoOnAllListeners.pluginJarAdded(DefaultPluginJarLocationMonitor.java:237)
        at com.thoughtworks.go.plugin.infra.monitor.PluginChangeNotifier.lambda$notify$0(PluginChangeNotifier.java:33)
        at java.base/java.lang.Iterable.forEach(Iterable.java:75)
        at com.thoughtworks.go.plugin.infra.monitor.PluginChangeNotifier.notify(PluginChangeNotifier.java:30)
        at com.thoughtworks.go.plugin.infra.monitor.DefaultPluginJarLocationMonitor$PluginLocationMonitorThread.loadAndNotifyPluginsFrom(DefaultPluginJarLocationMonitor.java:204)
        at com.thoughtworks.go.plugin.infra.monitor.DefaultPluginJarLocationMonitor$PluginLocationMonitorThread.oneShot(DefaultPluginJarLocationMonitor.java:192)
        at com.thoughtworks.go.plugin.infra.monitor.DefaultPluginJarLocationMonitor$PluginLocationMonitorThread.run(DefaultPluginJarLocationMonitor.java:177)
Caused by: java.lang.RuntimeException: java.nio.file.AccessDeniedException: /tmp/a
        at com.thoughtworks.go.util.ExceptionUtils.bomb(ExceptionUtils.java:37)
        at com.thoughtworks.go.server.service.plugins.AnalyticsPluginAssetsService.cacheStaticAssets(AnalyticsPluginAssetsService.java:137)
        at com.thoughtworks.go.server.service.plugins.AnalyticsPluginAssetsService.onPluginMetadataCreate(AnalyticsPluginAssetsService.java:87)
        at com.thoughtworks.go.plugin.access.analytics.AnalyticsMetadataLoader.pluginLoaded(AnalyticsMetadataLoader.java:50)
        at com.thoughtworks.go.plugin.infra.PluginLoader.lambda$doPostBundleInstallActivities$0(PluginLoader.java:113)
        at org.apache.commons.collections4.IteratorUtils.forEach(IteratorUtils.java:1224)
        at org.apache.commons.collections4.IterableUtils.forEach(IterableUtils.java:585)
        at com.thoughtworks.go.plugin.infra.PluginLoader.lambda$doPostBundleInstallActivities$1(PluginLoader.java:113)
        at org.apache.commons.collections4.IteratorUtils.forEach(IteratorUtils.java:1224)
        at org.apache.commons.collections4.IterableUtils.forEach(IterableUtils.java:585)
        at com.thoughtworks.go.plugin.infra.PluginLoader.doPostBundleInstallActivities(PluginLoader.java:112)
        at com.thoughtworks.go.plugin.infra.PluginLoader.loadPlugin(PluginLoader.java:70)
        ... 12 common frames omitted
Caused by: java.lang.RuntimeException: java.nio.file.AccessDeniedException: /tmp/a
        at com.thoughtworks.go.server.service.plugins.AnalyticsPluginAssetsService.safeCopyExternalAssetsToPluginAssetRoot(AnalyticsPluginAssetsService.java:159)
        at com.thoughtworks.go.server.service.plugins.AnalyticsPluginAssetsService.cacheStaticAssets(AnalyticsPluginAssetsService.java:131)
        ... 22 common frames omitted
Caused by: java.nio.file.AccessDeniedException: /tmp/a
        at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:90)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116)
        at java.base/sun.nio.fs.UnixFileSystemProvider.newDirectoryStream(UnixFileSystemProvider.java:412)
        at java.base/java.nio.file.Files.newDirectoryStream(Files.java:474)
        at java.base/java.nio.file.Files.list(Files.java:3769)
        at com.thoughtworks.go.server.service.plugins.AnalyticsPluginAssetsService.safeCopyExternalAssetsToPluginAssetRoot(AnalyticsPluginAssetsService.java:149)
        ... 23 common frames omitted
```